app/javascript/dashboard/helper/conversationHelper.js

```diff
@@ -0,0 +1,93 @@
+/**
+ * Determines the last non-activity message between store and API messages.
+ * @param {Object} messageInStore - The last non-activity message from the store.
+ * @param {Object} messageFromAPI - The last non-activity message from the API.
+ * @returns {Object} The latest non-activity message.
+ */
+const getLastNonActivityMessage = (messageInStore, messageFromAPI) => {
+  // If both API value and store value for last non activity message
+  // are available, then return the latest one.
+  if (messageInStore && messageFromAPI) {
+    return messageInStore.created_at >= messageFromAPI.created_at
+      ? messageInStore
+      : messageFromAPI;
+  }
+  // Otherwise, return whichever is available
+  return messageInStore || messageFromAPI;
+};
+
+/**
+ * Filters out duplicate source messages from an array of messages.
+ * @param {Array} messages - The array of messages to filter.
+ * @returns {Array} An array of messages without duplicates.
+ */
+export const filterDuplicateSourceMessages = (messages = []) => {
+  const messagesWithoutDuplicates = [];
+  // We cannot use Map or any short hand method as it returns the last message with the duplicate ID
+  // We should return the message with smaller id when there is a duplicate
+  messages.forEach(m1 => {
+    if (m1.source_id) {
+      if (
+        messagesWithoutDuplicates.findIndex(
+          m2 => m1.source_id === m2.source_id
+        ) < 0
+      ) {
+        messagesWithoutDuplicates.push(m1);
+      }
+    } else {
+      messagesWithoutDuplicates.push(m1);
+    }
+  });
+  return messagesWithoutDuplicates;
+};
+
+/**
+ * Retrieves the last message from a conversation, prioritizing non-activity messages.
+ * @param {Object} m - The conversation object containing messages.
+ * @returns {Object} The last message of the conversation.
+ */
+export const lastMessage = m => {
+  const lastMessageIncludingActivity = m.messages.at(-1);
+
+  const nonActivityMessages = m.messages.filter(
+    message => message.message_type !== 2
+  );
+  const lastNonActivityMessageInStore = nonActivityMessages.at(-1);
```

Comment:
Replaced `.last()` with `.at(-1)` to pick last non activity message. Faced some issue with specs. [Old code](https://github.com/chatwoot/chatwoot/pull/9931/files#diff-e296966f15bc9933602dd4154316927d6790dfe6bc39a7f3f113670dbef0c873L43). [Ref PR](https://github.com/chatwoot/chatwoot/pull/5864/files)
cc @pranavrajs @scmmishra